src/utils/body.ts

```diff
@@ -54,29 +72,92 @@ function convertFormDataToBodyData<T extends BodyData = BodyData>(
   formData: FormData,
   options: ParseBodyOptions
 ): T {
-  const form: BodyData = {}
+  const form: BodyData = Object.create(null)
 
   formData.forEach((value, key) => {
     const shouldParseAllValues = options.all || key.endsWith('[]')
 
     if (!shouldParseAllValues) {
-      form[key] = value
+      handleNestedValues(form, key, value, options.dot)
     } else {
-      handleParsingAllValues(form, key, value)
+      handleParsingAllValues(form, key, value, options.dot)
     }
   })
 
   return form as T
 }
 
-const handleParsingAllValues = (form: BodyData, key: string, value: FormDataEntryValue): void => {
-  const formKey = form[key]
+const handleParsingAllValues = (
+  form: BodyData,
+  key: string,
+  value: FormDataEntryValue,
+  dot?: boolean
+): void => {
+  if (dot && key.includes('.')) {
+    handleNestedValues(form, key, value, dot, true)
+  } else {
+    if (form[key] !== undefined) {
+      if (Array.isArray(form[key])) {
+        ;(form[key] as (string | File)[]).push(value)
+      } else {
+        form[key] = [form[key] as string | File, value]
+      }
+    } else {
+      form[key] = value
+    }
+  }
+}
 
-  if (formKey && Array.isArray(formKey)) {
-    ;(form[key] as (string | File)[]).push(value)
-  } else if (formKey) {
-    form[key] = [formKey, value]
+const handleNestedValues = (
```

Comment:
And finally here

```ts
const handleParsingNestedValues = (
  form: BodyData,
  key: string,
  value: FormDataEntryValue,
  parseAllValues: boolean
): void => {
  let nestedForm = form
  const keys = key.split('.')

  keys.forEach((key, index) => {
    if (index === keys.length - 1) {
      if (parseAllValues) {
        handleParsingAllValues(nestedForm, key, value)
      } else {
        nestedForm[key] = value
      }
    } else {
      if (
        !nestedForm[key] ||
        typeof nestedForm[key] !== 'object' ||
        Array.isArray(nestedForm[key]) ||
        nestedForm[key] instanceof File
      ) {
        nestedForm[key] = Object.create(null)
      }
      nestedForm = nestedForm[key] as BodyData
    }
  })
}
```

We can improve this more by removing the `parseAllValues` param in this function and somehow abstracting the logic to assign the value back in the `convertFormDataToBodyData` loop or maybe something else. I didn't want to add a dynamic function to manage this, but it's up to you guys, we can go that way also

Reply:
I think creating a new function to just handle nested on all values is good to avoid a dynamic function, what do you think?

```ts
  formData.forEach((value, key) => {
    const shouldParseAllValues = options.all || key.endsWith('[]')
    const shouldParseDotValues = options.dot && key.includes('.')

    if (shouldParseDotValues) {
      if (shouldParseAllValues) {
        handleParsingNestedAllValues(form, key, value)
      } else {
        handleParsingNestedValues(form, key, value)
      }
    } else {
      assignValue(form, key, value, shouldParseAllValues)
    }
  })
```

Reply:
This will just add bloat rather than help. Honestly, the functionality is working so this is just over-engineering. Let me explain the problem I am seeing in the current code. Down below I have shown the 2 worst case scenarios for this
```js
FormData {
  "obj.some.thing": "1",
  "obj.some.thing": "1",
  "obj.some.thing": "1",
}

// OR

FormData {
  "obj.some.thing": "1",
  "obj.some.one": "1",
  "obj.some.hello": "1",
}
```

Both cases have a complexity of **O(mn)** where _m_ is the number of entries and _n_ is the number of subkeys in an entry. We can improve this to **O(mn)** simply by changing the loop of `convertFormDataToBodyData` and minor changes in the `handleParsingNestedValues` function. Here are the changes respectively -

We are just processing the values separately. This first put the data in arrays the same as the old code and then converts it into a nested structure if the dot prop is `true`, only to the ones who have a `.`.
```ts
const form: BodyData = Object.create(null)

formData.forEach((value, key) => {
  const shouldParseAllValues = options.all || key.endsWith('[]')

  if (shouldParseAllValues) {
    handleParsingAllValues(form, key, value)
  } else {
    form[key] = value
  }
})

if (options.dot) {
  const nestedForm: BodyData = Object.create(null)

  Object.entries(form).forEach(([key, value]) => {
    const shouldParseDotValues = key.includes('.')

    if (shouldParseDotValues) {
      handleParsingNestedValues(nestedForm, key, value)
    } else {
      nestedForm[key] = value
    }
  })

  return nestedForm as T
}

return form as T
```
And remove the `parseAllValues` param as we don't need it now.
```ts
const handleParsingNestedValues = (
  form: BodyData,
  key: string,
  value: BodyDataValue
): void => {
  let nestedForm = form
  const keys = key.split('.')

  keys.forEach((key, index) => {
    if (index === keys.length - 1) {
        nestedForm[key] = value
    } else {
      if (
        !nestedForm[key] ||
        typeof nestedForm[key] !== 'object' ||
        Array.isArray(nestedForm[key]) ||
        nestedForm[key] instanceof File
      ) {
        nestedForm[key] = Object.create(null)
      }
      nestedForm = nestedForm[key] as BodyData
    }
  })
}
```
Overall this will improve the simplicity of the code and also the complexity. We can further improve the complexity of this but as I said before it's just over-engineering. Until @yusukebe wants us to improve this further we are good.

Reply:
@fzn0x, just some last-minute formatting of `convertFormDataToBodyData` function, wrote it without thinking so much. Thanks

```ts
function convertFormDataToBodyData<T extends BodyData = BodyData>(
  formData: FormData,
  options: ParseBodyOptions
): T {
  const form: BodyData = Object.create(null)

  formData.forEach((value, key) => {
    const shouldParseAllValues = options.all || key.endsWith('[]')

    if (!shouldParseAllValues) {
      form[key] = value
    } else {
      handleParsingAllValues(form, key, value)
    }
  })

  if (options.dot) {
    Object.entries(form).forEach(([key, value]) => {
      const shouldParseDotValues = key.includes('.')

      if (shouldParseDotValues) {
        handleParsingNestedValues(form, key, value)
        delete form[key]
      }
    })
  }

  return form as T
}
```